csrc/device_lower/pass/alias_memory.cpp

```diff
@@ -1385,62 +1409,370 @@ class AllocationReuseModifier : private kir::ExprMutator {
+//! pop as soon as possible. The algorithm would do the following in the above
+//! example:
+//!
+//!   a: Append A to holding area
+//!   b: Append B to holding area
+//!   c: Sort holding area by last read (desc): {B, A}.
+//!      Push B.
+//!      Push A.
+//!      Pop A.
+//!      Clear holding area.
+//!   d: Allocate C on top of B. RE-USES A
+//!   e: Cannot pop B since it is covered by C
+//!   f: Pop C and B
+//!
+//!   +-----+ +-----+
+//!   |  A  | |  C  |
+//!   +---+---+-+---+
+//!       |  B  |
+//!       +-+---+
+//!   a   b c d e   f
+//!
+//! Note that this only gives one layer of protection. More complex patterns can
+//! still result in suboptimal memory use even with reordering using the holding
+//! area.
+//!
+//! [Syncs for Reused (Not Aliased) Shared Memory]
+//! We will need to ensure the block is synced to prevent a race hazard where
+//! C is written to before some threads have read B, leading to corruption of
+//! those threads. These syncs need to ensure that the last read that might
+//! overlap the new allocation has finished before the first write of the new
+//! allocation. In this case that could be accomplished with an arrive/wait
+//! barrier between time points c and d, but any synchronizations contained
+//! within that interval in the original kernel, including a __syncthreads(),
+//! should be preferred over inserting new syncs.
+//!
+//! [Aliased Allocations]
+//! We handle aliased allocations by using a single liveness interval spanning
+//! from the first write of the base Allocation to the last read of either that
+//! allocation or the last alias. In some cases, this could be suboptimal; for
+//! example if a large tensor is used briefly at the beginning and is aliased
+//! near the end of the Fusion, then with this approach we will not be able to
+//! re-use its memory in the middle, which might be wasteful.
+class StackBasedSharedMemAllocator : kir::IrVisitor {
  public:
+  StackBasedSharedMemAllocator(const AllocationInfoMap& allocation_info_map)
+      : allocation_info_map_(allocation_info_map) {}
+
   void allocate(std::vector<Expr*>& exprs) {
+    recordEvents();
+
+    // Traverse expressions: reclaim memory when we pass a blockSync, append to
+    // waiting_to_push_ when we pass an Allocate
     handle(exprs);
+
+    // This is done whenever we pass a syncing op, but we need to do it again in
+    // case there are some allocations waiting around to be allocated.
+    sortPushAndAssignWaiting();
   }
 
  private:
   using kir::IrVisitor::handle;
 
-  static Val* alignExpr(Val* addr, int64_t alignment = 16) {
-    if (alignment == 1) {
-      return addr;
+  void dispatch(Expr* expr) final {
+    auto pos_opt = allocation_info_map_.getScopeMap().getMaybeExprPos(expr);
+    if (!pos_opt.has_value()) {
+      // Aliased kir::Allocate exprs are replaced, so the replacements will not
+      // exist in the original ScopeMap. We handle those manually here.
+      position_++;
```

Comment:
Should we just make `AllocationReuseModifier` update this map?

Reply:
Good idea.